## Import Libraries


In [23]:
#import the packages
import pandas as pd
import pandas_profiling
import numpy as np
import matplotlib
import matplotlib.pyplot as plt
import seaborn as sns
import tkinter
matplotlib.use('TkAgg')


## Read the Data

In [16]:
# read the file
df = pd.read_csv('C:\\HD\\Use_case_data_set.csv', index_col=0, parse_dates=True)
df.dtypes

price            object
mpg               int64
rep78            object
headroom          int64
trunk             int64
weight           object
length            int64
turn              int64
displacement      int64
gear_ratio      float64
foreign          object
dtype: object

## Checking the Data Distribution


In [17]:
df.describe(include="all")

price        mpg    rep78   headroom      trunk weight      length  \
count      74  74.000000       69  74.000000  74.000000     74   74.000000   
unique     74        NaN        5        NaN        NaN     64         NaN   
top     3,895        NaN  Average        NaN        NaN  3,690         NaN   
freq        1        NaN       30        NaN        NaN      2         NaN   
mean      NaN  21.297297      NaN   3.243243  13.756757    NaN  187.932432   
std       NaN   5.785503      NaN   0.888447   4.277404    NaN   22.266340   
min       NaN  12.000000      NaN   2.000000   5.000000    NaN  142.000000   
25%       NaN  18.000000      NaN   3.000000  10.250000    NaN  170.000000   
50%       NaN  20.000000      NaN   3.000000  14.000000    NaN  192.500000   
75%       NaN  24.750000      NaN   4.000000  16.750000    NaN  203.750000   
max       NaN  41.000000      NaN   5.000000  23.000000    NaN  233.000000   

             turn  displacement  gear_ratio   foreign  
count   74.000000     74.000000   74.000000        74  
unique        NaN           NaN         NaN         2  
top           NaN           NaN         NaN  Domestic  
freq          NaN           NaN         NaN        52  
mean    39.648649    197.297297    3.014865       NaN  
std      4.399354     91.837219    0.456287       NaN  
min     31.000000     79.000000    2.190000       NaN  
25%     36.000000    119.000000    2.730000       NaN  
50%     40.000000    196.000000    2.955000       NaN  
75%     43.000000    245.250000    3.352500       NaN  
max     51.000000    425.000000    3.890000       NaN

## Cleanning the Data 

Here data type to the price feature is “object” the expected data type should really be an integer or float type.
To covert datatype: dataframe.astype()

### Check Null Values

In [18]:
df.isnull().sum()

price           0
mpg             0
rep78           5
headroom        0
trunk           0
weight          0
length          0
turn            0
displacement    0
gear_ratio      0
foreign         0
dtype: int64

## Check Symbols

In [19]:
for col in df.columns:
    print('{} : {}'.format(col,df[col].unique()))

price : ['13,594' '11,497' '14,500' '13,466' '6,303' '11,385' '12,990' '5,379'
 '7,827' '6,165' '10,371' '5,886' '10,372' '5,705' '9,690' '4,749' '6,342'
 '11,995' '5,798' '4,060' '5,899' '4,890' '5,788' '4,934' '4,010' '4,516'
 '5,719' '5,222' '4,181' '4,723' '4,424' '4,082' '5,172' '3,955' '4,733'
 '4,816' '3,291' '5,189' '8,129' '4,296' '8,814' '15,906' '4,187' '3,829'
 '5,104' '4,504' '3,799' '4,099' '6,229' '7,140' '6,295' '4,172' '4,195'
 '5,079' '3,667' '5,799' '4,697' '4,482' '9,735' '6,850' '4,453' '6,486'
 '3,895' '4,647' '4,389' '4,499' '3,299' '3,995' '3,984' '3,748' '4,425'
 '4,589' '3,798' '5,397']
mpg : [12 14 15 16 17 18 19 20 21 22 23 24 25 26 28 29 30 31 34 35 41]
rep78 : ['Average' 'Fair' 'Good' nan 'Excellent' 'Poor']
headroom : [3 4 5 2]
trunk : [18 22 16 15 20 14 23 17 11 21  7 13  8 10  9 12  6  5]
weight : ['4,720' '4,840' '3,900' '3,830' '4,130' '4,330' '3,420' '4,060' '4,080'
 '3,720' '4,030' '3,600' '3,880' '3,690' '2,830' '3,350' '3,740' '3,170'
 '3,700' '3,

### Making Data Readable
Fill the NaN with Zeros

In [20]:
df[['mpg']]=df['mpg'].replace(',', '').astype(float)
df[['foreign']]=df['foreign'].replace('Foreign', '1')
df[['foreign']]=df['foreign'].replace('Domestic', '0')
df[['rep78']]=df['rep78'].replace('Poor', '0')
df[['rep78']]=df['rep78'].replace('Fair', '1')
df[['rep78']]=df['rep78'].replace('Average', '2')
df[['rep78']]=df['rep78'].replace('Good', '3')
df[['rep78']]=df['rep78'].replace('Excellent', '4')
df.fillna(0)

df

price   mpg rep78  headroom  trunk weight  length  turn  \
make                                                                          
Linc. Mark V       13,594  12.0     2         3     18  4,720     230    48   
Linc. Continental  11,497  12.0     2         4     22  4,840     233    51   
Cad. Eldorado      14,500  14.0     1         4     16  3,900     204    43   
Linc. Versailles   13,466  14.0     2         4     15  3,830     201    41   
Merc. XR-7          6,303  14.0     3         3     16  4,130     217    45   
...                   ...   ...   ...       ...    ...    ...     ...   ...   
Toyota Corolla      3,748  31.0     4         3      9  2,200     165    35   
Plym. Champ         4,425  34.0     4         3     11  1,800     157    37   
Datsun 210          4,589  35.0     4         2      8  2,020     165    32   
Subaru              3,798  35.0     4         3     11  2,050     164    36   
VW Diesel           5,397  41.0     4         3     15  2,040     155    35   

                   displacement  gear_ratio foreign  
make                                                 
Linc. Mark V                400        2.47       0  
Linc. Continental           400        2.47       0  
Cad. Eldorado               350        2.19       0  
Linc. Versailles            302        2.47       0  
Merc. XR-7                  302        2.75       0  
...                         ...         ...     ...  
Toyota Corolla               97        3.21       1  
Plym. Champ                  86        2.97       0  
Datsun 210                   85        3.70       1  
Subaru                       97        3.81       1  
VW Diesel                    90        3.78       1  

[74 rows x 11 columns]

### Visualizing the missing values

In [24]:

sns.heatmap(df.isnull(),cbar=False,cmap='viridis')
plt.show()

### Replace Comma's and Change the data types

In [6]:
df.replace(',','', regex=True, inplace=True)
c = df.select_dtypes(object).columns
df[c] = df[c].apply(pd.to_numeric,errors='coerce')


#### Review Data

In [70]:
df.info

<bound method DataFrame.info of                    price   mpg  rep78  headroom  trunk  weight  length  turn  \
make                                                                           
Linc. Mark V       13594  12.0    2.0         3     18    4720     230    48   
Linc. Continental  11497  12.0    2.0         4     22    4840     233    51   
Cad. Eldorado      14500  14.0    1.0         4     16    3900     204    43   
Linc. Versailles   13466  14.0    2.0         4     15    3830     201    41   
Merc. XR-7          6303  14.0    3.0         3     16    4130     217    45   
...                  ...   ...    ...       ...    ...     ...     ...   ...   
Toyota Corolla      3748  31.0    4.0         3      9    2200     165    35   
Plym. Champ         4425  34.0    4.0         3     11    1800     157    37   
Datsun 210          4589  35.0    4.0         2      8    2020     165    32   
Subaru              3798  35.0    4.0         3     11    2050     164    36   
VW Diese

### Generate Data Profile

In [8]:
# read the file
#df = pd.read_csv('C:\\HD\\Use_case_data_set.csv', index_col=0, parse_dates=True)
   
# run the profile report
profile = df.profile_report(title='Pandas Profiling Report')
   
# save the report as html file
profile.to_file(output_file="pandas_profiling1.html")
   
# save the report as json file
profile.to_file(output_file="pandas_profiling2.json")

Export report to file: 100%|██████████| 1/1 [00:00<00:00, 143.14it/s]


### Identify the Outliers in the dataset with numeric columns

In [9]:
sns.boxplot(x="mpg",y="gear_ratio",data=df)
plt.show()

<ipython-input-9-f98dc58804ec>:2: UserWarning: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.
  plt.show()


### Mileage vs price

In [74]:
sns.boxplot(x="mpg",y="price",data=df)
plt.show()

#### Gear Ration Vs mpg

In [17]:
sns.boxplot(x="gear_ratio",y="mpg",data=df)
plt.show()

Just taking a quick glance of our boxplot, we can see that the average gear_ratio with respect to mpg

### Pearson Correlation
The Pearson Correlation measures the linear dependence between two variables X and Y. The resulting coefficient is a value between -1 and 1 inclusive, where:
1: total positive linear correlation,
0: no linear correlation, the two variables most likely do not affect each other
-1: total negative linear correlation.

In [77]:
df.corr()

price       mpg     rep78  headroom     trunk    weight  \
price         1.000000 -0.468597  0.006553  0.161288  0.314332  0.538611   
mpg          -0.468597  1.000000  0.402340 -0.424679 -0.581585 -0.807175   
rep78         0.006553  0.402340  1.000000 -0.147239 -0.157243 -0.400344   
headroom      0.161288 -0.424679 -0.147239  1.000000  0.610553  0.462438   
trunk         0.314332 -0.581585 -0.157243  0.610553  1.000000  0.672206   
weight        0.538611 -0.807175 -0.400344  0.462438  0.672206  1.000000   
length        0.431831 -0.795779 -0.360565  0.501493  0.726596  0.946009   
turn          0.309617 -0.719186 -0.496131  0.393669  0.601059  0.857443   
displacement  0.494943 -0.705643 -0.411943  0.445691  0.608635  0.894896   
gear_ratio   -0.313661  0.616177  0.410340 -0.311138 -0.508665 -0.759258   
foreign       0.048719  0.393397  0.592237 -0.246307 -0.359435 -0.592830   

                length      turn  displacement  gear_ratio   foreign  
price         0.431831  0.309617      0.494943   -0.313661  0.048719  
mpg          -0.795779 -0.719186     -0.705643    0.616177  0.393397  
rep78        -0.360565 -0.496131     -0.411943    0.410340  0.592237  
headroom      0.501493  0.393669      0.445691   -0.311138 -0.246307  
trunk         0.726596  0.601059      0.608635   -0.508665 -0.359435  
weight        0.946009  0.857443      0.894896   -0.759258 -0.592830  
length        1.000000  0.864261      0.835140   -0.696383 -0.570198  
turn          0.864261  1.000000      0.776765   -0.676300 -0.631096  
displacement  0.835140  0.776765      1.000000   -0.828877 -0.613758  
gear_ratio   -0.696383 -0.676300     -0.828877    1.000000  0.706726  
foreign      -0.570198 -0.631096     -0.613758    0.706726  1.000000

In [81]:
y=df['price']
x=df['mpg']
plt.bar(x,y)

plt.show() 

In [84]:
plt.scatter(df['mpg'],df['rep78'],color='b')
plt.xlabel('mpg')
plt.ylabel('rep78')
plt.show()

### Hence the spearmans calculation are right - there is a corelations between gear_ratio and mpg

### Exploratory Data Analysis 
Are we asking the right questions
Distribution, etc

In [11]:
plt.figure(figsize=(10,10))
sns.heatmap(df.corr(),cbar=True,annot=True,cmap='Blues')
plt.show()

<ipython-input-11-7614cfd7d923>:3: UserWarning: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.
  plt.show()


### Positive Correlation
Price – weight, mpg, rep78 
Negative Correlation
Price – turn, displacement
This heatmap has given us great insights into the data.

#### How does the gear_ratio affect the price?

In [10]:
plt.figure(figsize=(10,10))
plt.scatter(x='gear_ratio',y='price',data=df)
plt.xlabel('gear_ratio')
plt.ylabel('Price')
plt.show()

<ipython-input-10-be5dea3063ff>:5: UserWarning: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.
  plt.show()


##### We can see that most of the gear _ ratio value lies between 2.75 - 3.25 has price mostly between 4000-60000, there are outliers also (between 10000-16000).

### Let us check a count between 2.75 - 3.25 i.e univariate analysis of gear_ratio.

In [32]:
sns.histplot(df.gear_ratio,bins=10)
plt.show()

The average count between 2.75 - 3.25 is 12 and it is positively skewed.

##### Checking an relationships and impact of vairables

In [33]:
for col in df.columns:
    sns.scatterplot(x=col,y='mpg', data=df)
    plt.show()

##### How does mpg affect price

In [38]:
plt.figure(figsize=(10,10))
plt.scatter(x='mpg',y='price',data=df)
plt.xlabel('mpg')
plt.ylabel('Price')
plt.show()

We can see price decreases with an increase in mpg.

## Using the Corrrelation now to find the relations

In [ ]:
df.corr('spearman')

In [ ]:
df.corr('kendall')

## Predicting
What feature can be used to create a model from this data
